File: Complementar - Estruturas de dados (Didático)/Algoritmos21/Alg21.py
'''
♦ Algoritmo 21: Estrutura de Dados Não Linear Estática (não Dinâmica) de Árvore Binária - Busca Binária (Binary Search Tree - BST) de tipo de dados genérico.
→ Operações:
    • Inserção e Remoção de Elementos;
    • Busca (pesquisa) de Elementos;
    • Encontrar o Mínimo e o Máximo;
    • Travessias (in-order, pre-order, post-order);
    • Verificar se é uma BST válida.
'''

class BinarySearchTree:
    class TreeNode:
        def __init__(self, data):
            self.data = data
            self.left = None
            self.right = None

    def __init__(self):
        self.root = None

    # Inserção de elementos
    def insert(self, data):
        self.root = self.insertRec(self.root, data)

    def insertRec(self, root, data):
        if root is None:
            return self.TreeNode(data)

        if data < root.data:
            root.left = self.insertRec(root.left, data)
        elif data > root.data:
            root.right = self.insertRec(root.right, data)

        return root

    # Remoção de elementos
    def delete(self, data):
        self.root = self.deleteRec(self.root, data)

    def deleteRec(self, root, data):
        if root is None:
            return root

        if data < root.data:
            root.left = self.deleteRec(root.left, data)
        elif data > root.data:
            root.right = self.deleteRec(root.right, data)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left

            root.data = self.minValue(root.right)

            root.right = self.deleteRec(root.right, root.data)

        return root

    def minValue(self, root):
        min_value = root.data
        while root.left is not None:
            min_value = root.left.data
            root = root.left
        return min_value

    # Busca de elementos
    def search(self, data):
        return self.searchRec(self.root, data)

    def searchRec(self, root, data):
        if root is None:
            return False

        if root.data == data:
            return True

        if data < root.data:
            return self.searchRec(root.left, data)

        return self.searchRec(root.right, data)

    # Encontrar o Mínimo e o Máximo
    def findMin(self):
        if self.root is None:
            return None

        current = self.root
        while current.left is not None:
            current = current.left

        return current.data

    def findMax(self):
        if self.root is None:
            return None

        current = self.root
        while current.right is not None:
            current = current.right

        return current.data

    # Travessias (in-order, pre-order, post-order)
    def inOrder(self):
        self.inOrderRec(self.root)

    def inOrderRec(self, root):
        if root is not None:
            self.inOrderRec(root.left)
            print(root.data, end=" ")
            self.inOrderRec(root.right)

    def preOrder(self):
        self.preOrderRec(self.root)

    def preOrderRec(self, root):
        if root is not None:
            print(root.data, end=" ")
            self.preOrderRec(root.left)
            self.preOrderRec(root.right)

    def postOrder(self):
        self.postOrderRec(self.root)

    def postOrderRec(self, root):
        if root is not None:
            self.postOrderRec(root.left)
            self.postOrderRec(root.right)
            print(root.data, end=" ")

    # Verificar se é uma BST válida
    def isValidBST(self):
        return self.isValidBSTRec(self.root, None, None)

    def isValidBSTRec(self, root, min_value, max_value):
        if root is None:
            return True

        if (min_value is not None and root.data <= min_value) or (max_value is not None and root.data >= max_value):
            return False

        return self.isValidBSTRec(root.left, min_value, root.data) and self.isValidBSTRec(root.right, root.data, max_value)


tree = BinarySearchTree()
tree.insert(50)
tree.insert(30)
tree.insert(20)
tree.insert(40)
tree.insert(70)
tree.insert(60)
tree.insert(80)

print("In-order traversal:")
tree.inOrder()
print()

print("Pre-order traversal:")
tree.preOrder()
print()

print("Post-order traversal:")
tree.postOrder()
print()

print("Minimum value:", tree.findMin())
print("Maximum value:", tree.findMax())

searchValue = 40
print("Search for", searchValue, ":", tree.search(searchValue))

tree.delete(searchValue)
print("After deleting", searchValue, ":")
tree.inOrder()

print("\nIs valid BST?", tree.isValidBST())

File: Complementar - Estruturas de dados (Didático)/Algoritmos21/Alg21.cpp
/*
♦ Algoritmo 21: Estrutura de Dados Não Linear Estática (não Dinâmica) de Árvore Binária - Busca Binária (Binary Search Tree - BST) de tipo de dados genérico.
→ Operações:
    • Inserção e Remoção de Elementos;
    • Busca (pesquisa) de Elementos;
    • Encontrar o Mínimo e o Máximo;
    • Travessias (in-order, pre-order, post-order);
    • Verificar se é uma BST válida.
*/

#include <iostream>

template <typename T>
class BinarySearchTree {
    struct TreeNode {
        T data;
        TreeNode* left;
        TreeNode* right;

        TreeNode(T data) : data(data), left(nullptr), right(nullptr) {}
    };

    TreeNode* root;

public:
    BinarySearchTree() : root(nullptr) {}

    // Inserção de elementos
    void insert(const T& data) {
        root = insertRec(root, data);
    }

    TreeNode* insertRec(TreeNode* root, const T& data) {
        if (root == nullptr) {
            return new TreeNode(data);
        }

        if (data < root->data) {
            root->left = insertRec(root->left, data);
        } else if (data > root->data) {
            root->right = insertRec(root->right, data);
        }

        return root;
    }

    // Remoção de elementos
    void remove(const T& data) {
        root = removeRec(root, data);
    }

    TreeNode* removeRec(TreeNode* root, const T& data) {
        if (root == nullptr) {
            return root;
        }

        if (data < root->data) {
            root->left = removeRec(root->left, data);
        } else if (data > root->data) {
            root->right = removeRec(root->right, data);
        } else {
            if (root->left == nullptr) {
                TreeNode* temp = root->right;
                delete root;
                return temp;
            } else if (root->right == nullptr) {
                TreeNode* temp = root->left;
                delete root;
                return temp;
            }

            root->data = minValue(root->right);

            root->right = removeRec(root->right, root->data);
        }

        return root;
    }

    T minValue(TreeNode* root) {
        T minValue = root->data;
        while (root->left != nullptr) {
            minValue = root->left->data;
            root = root->left;
        }
        return minValue;
    }

    // Busca de elementos
    bool search(const T& data) {
        return searchRec(root, data);
    }

    bool searchRec(TreeNode* root, const T& data) {
        if (root == nullptr) {
            return false;
        }

        if (root->data == data) {
            return true;
        }

        if (data < root->data) {
            return searchRec(root->left, data);
        }

        return searchRec(root->right, data);
    }

    // Encontrar o Mínimo e o Máximo
    T findMin() {
        if (root == nullptr) {
            return T(); // Retorna um valor padrão para indicar que a árvore está vazia
        }

        TreeNode* current = root;
        while (current->left != nullptr) {
            current = current->left;
        }

        return current->data;
    }

    T findMax() {
        if (root == nullptr) {
            return T(); // Retorna um valor padrão para indicar que a árvore está vazia
        }

        TreeNode* current = root;
        while (current->right != nullptr) {
            current = current->right;
        }

        return current->data;
    }

    // Travessias (in-order, pre-order, post-order)
    void inOrder() {
        inOrderRec(root);
    }

    void inOrderRec(TreeNode* root) {
        if (root != nullptr) {
            inOrderRec(root->left);
            std::cout << root->data << " ";
            inOrderRec(root->right);
        }
    }

    void preOrder() {
        preOrderRec(root);
    }

    void preOrderRec(TreeNode* root) {
        if (root != nullptr) {
            std::cout << root->data << " ";
            preOrderRec(root->left);
            preOrderRec(root->right);
        }
    }

    void postOrder() {
        postOrderRec(root);
    }

    void postOrderRec(TreeNode* root) {
        if (root != nullptr) {
            postOrderRec(root->left);
            postOrderRec(root->right);
            std::cout << root->data << " ";
        }
    }

    // Verificar se é uma BST válida
    bool isValidBST() {
        return isValidBSTRec(root, nullptr, nullptr);
    }

    bool isValidBSTRec(TreeNode* root, const T* min, const T* max) {
        if (root == nullptr) {
            return true;
        }

        if ((min != nullptr && root->data <= *min) || (max != nullptr && root->data >= *max)) {
            return false;
        }

        return isValidBSTRec(root->left, min, &(root->data)) && isValidBSTRec(root->right, &(root->data), max);
    }

    ~BinarySearchTree() {
        destroyTree(root);
    }

private:
    void destroyTree(TreeNode* root) {
        if (root != nullptr) {
            destroyTree(root->left);
            destroyTree(root->right);
            delete root;
        }
    }
};

int main() {
    BinarySearchTree<int> tree;
    tree.insert(50);
    tree.insert(30);
    tree.insert(20);
    tree.insert(40);
    tree.insert(70);
    tree.insert(60);
    tree.insert(80);

    std::cout << "In-order traversal:" << std::endl;
    tree.inOrder();
    std::cout << std::endl;

    std::cout << "Pre-order traversal:" << std::endl;
    tree.preOrder();
    std::cout << std::endl;

    std::cout << "Post-order traversal:" << std::endl;
    tree.postOrder();
    std::cout << std::endl;

    std::cout << "Minimum value: " << tree.findMin() << std::endl;
    std::cout << "Maximum value: " << tree.findMax() << std::endl;

    int searchValue = 40;
    std::cout << "Search for " << searchValue << ": " << tree.search(searchValue) << std::endl;

    tree.remove(40);
    std::cout << "After deleting " << searchValue << ":" << std::endl;
    tree.inOrder();

    std::cout << "\nIs valid BST? " << tree.isValidBST() << std::endl;

    return 0;
}
